src/App.tsx

```diff
@@ -14,11 +16,206 @@ const goodsFromServer: string[] = [
   'Garlic',
 ];
 
-const App: React.FC = () => (
-  <div className="App">
-    <h1>Goods</h1>
-    {goodsFromServer.length}
-  </div>
-);
+interface State {
+  showGoods: boolean,
+  goods: string[],
+  isReversed: boolean,
+  isSorted: boolean,
+  sortBy: string,
+  minLength: number,
+}
+
+class App extends React.Component<{}, State> {
+  state: Readonly<State> = {
+    showGoods: false,
+    goods: goodsFromServer,
+    isReversed: false,
+    isSorted: false,
+    sortBy: '',
+    minLength: 1,
+  };
+
+  handleGoodsList = () => {
+    this.setState({ showGoods: true });
+  };
+
+  handleSortAlphabetically = () => {
+    this.setState({ isSorted: true, sortBy: 'alphabet' });
+  };
+
+  handleSortByLength = () => {
+    this.setState({ isSorted: true, sortBy: 'length' });
+  };
+
+  handleReset = () => {
+    this.setState({
+      isReversed: false,
+      isSorted: false,
+      sortBy: '',
+      minLength: 1,
+    });
+  };
+
+  handleReverse = () => {
+    this.setState(state => ({
+      isReversed: !state.isReversed,
+    }));
+  };
+
+  handleMinLength = (value: number) => {
+    this.setState({ minLength: value });
+  };
+
+  render() {
+    const {
+      showGoods,
+      goods,
+      isReversed,
+      isSorted,
+      sortBy,
+      minLength,
+    } = this.state;
+
+    const {
+      handleGoodsList,
+      handleReverse,
+      handleSortAlphabetically,
+      handleSortByLength,
+      handleReset,
+      handleMinLength,
+    } = this;
+
+    const visibleGoods = goods.filter(
+      singleGoods => singleGoods.length >= minLength,
+    );
+
+    if (isSorted) {
+      switch (sortBy) {
+        case 'alphabet':
+          visibleGoods.sort((firstProduct, secondProduct) => {
+            return firstProduct.localeCompare(secondProduct);
+          });
+          break;
+
+        case 'length':
+          visibleGoods.sort((firstProduct, secondProduct) => {
+            return firstProduct.length - secondProduct.length;
+          });
+          break;
+
+        default:
+          return 0;
+      }
+    }
+
+    if (isReversed) {
+      visibleGoods.reverse();
+    }
```

Comment:
This logic shouldn't be here. Method render should be as clear as possible.
Let's move this logic to a separate method